### PR TITLE
Prune imported indexes that are not used by the dataflow

### LINKED ISSUE #4887: Restrict arrangement imports to those we use
The dataflow layer currently imports all arrangements for a source, because we don't have great communication about which arrangements we will actually use. We roughly know the points where arrangements *are* used, as we perform analyses, but this information is not communicated back up.

While not a major problem, it creates dataflow graphs that look like
Screen Shot 2020-11-26 at 7 20 25 AM

### PR DESCRIPTION
**Pull request by asenac**
This patch adds an extra optimization step after view inlining that removes all imported indexes that won't be used by the dataflow.

I find this a bit fragile since this separate piece of logic must be kept in sync with the rendering code that resolves the indexes later. My personal preference would be to modify the dataflow to make explicit what indexes are used for each arrangement.

The test uses a similar query to the one that is used by the memory profiler tool.

Note: I've used a testdrive test since sqllogic runs in symbiosis mode and fails when accessing `mz_catalog` tables

Fixes #4887.

### CODE REVIEW

File: src/transform/src/dataflow.rs [hunk 1]
```diff
@@ -167,6 +167,72 @@ fn inline_views(dataflow: &mut DataflowDesc) {
...
+/// Finds the imported indexes that are actually used by any arrangement in the
+/// dataflow and remove the rest from the collection of imported ones.
+fn prune_unused_imported_indexes(
+    dataflow: &mut DataflowDesc,
+    indexes_by_collection: &HashMap<GlobalId, Vec<(GlobalId, Vec<MirScalarExpr>)>>,
+) {
+    // Traverse the dataflow to find the imported indexes that are actually used
+    let mut used_indexes = HashSet::new();
+    for object in dataflow.objects_to_build.iter_mut() {
+        object.relation_expr.as_mut().visit_mut_pre(&mut |x| {
```
Comment (philip-stoev) [inline src/transform/src/dataflow.rs:183]: Totally random question, since you are not mutating the object as you walk it, do you need all the "mut" stuff?

Comment (asenac) [reply src/transform/src/dataflow.rs:183]: I guess I could use a different non-mutable visitor method.

One thing I'd like thought, and I haven't found, is a walker method that allows you to prune part of the subgraph. For example, once we have matched the current node against the nested pattern `ArrangeBy -> Get` we don't need to visit the nested `Get` operator.

File: test/testdrive/render-index-import.td [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+> create index i22 on t2 (f2);
+> create index i23 on t2 (f3, f2);
+> create table t3 (f1 integer, f2 integer, f3 integer);
+> create index i31 on t3 (f1);
+> create index i32 on t3 (f2);
+> create index i33 on t3 (f3, f2);
+
+# All tables are joined on column 1
+> create materialized view v1 as select *  from t1, t2, t3 where t1.f2=t2.f2 AND t1.f2 = t3.f2;
+> SELECT DISTINCT (SELECT name FROM mz_catalog.mz_tables WHERE id = parts[1]) AS tablename, parts[2] AS columns FROM (SELECT regexp_match(name, 'Index\((u\d+), (.*)\)') AS parts FROM mz_catalog.mz_dataflow_operators WHERE id IN (SELECT id FROM mz_catalog.mz_dataflow_operator_addresses WHERE address[1] = (SELECT DISTINCT address[1] FROM mz_catalog.mz_dataflow_operator_addresses WHERE id = (SELECT DISTINCT id from mz_catalog.mz_dataflow_names WHERE name LIKE '%.v1%'))) AND name LIKE 'Index%') ORDER BY tablename, columns;
```
Comment (asenac) [inline test/testdrive/render-index-import.td:27]: `testdrive` doesn't seem to support multiline queries, does it?

Comment (philip-stoev) [reply test/testdrive/render-index-import.td:27]: It does, however the second and every subsequent line needs to be indented 2 spaces from the left:

```
> SELECT something
  FROM something
```

Comment (asenac) [reply test/testdrive/render-index-import.td:27]: oh, I see. I'll beautify the test then :) Thanks!

**Review by philip-stoev — APPROVED** [review]
Thank you for the tests!

I visually checked the graphs for a bunch of TPC-like queries and there are no unconnected Index or Map nodes.

No crashes on the TPC-like randomized workload.

File: src/transform/src/dataflow.rs (hunk 1 above)
Comment (wangandi) [reply src/transform/src/dataflow.rs:183]: In `src/expr/src/scalar/mod.rs` (a.k.a. the implementation for `MirScalarExpr`, there is a method `visit_mut_pre_post` that allows you specify whether or not you want to keep recursing down. See if adding similar method to `MirRelationExpr` would work.

**Comment by asenac** [comment]
Closing PR due to https://github.com/MaterializeInc/materialize/issues/4887#issuecomment-845278946.

**Pull request closed without merging**